docs/rules/array-foreach.md

```diff
@@ -14,6 +14,60 @@ for (const el of els) {
 }
 ```
 
+## Why disallow `forEach`
+
+Here's a summary of why `forEach` is disallowed, and why we prefer `for...of` for almost any use-case of `forEach`:
+
+ - Allowing `forEach` encourages **layering of "bad practices"**, such as using `Array.from()` (which is less performant than using `for...of`).
+ - When more requirements are added on, `forEach` typically gets **chained** with other methods like `filter` or `map`, causing multiple iterations over the same Array. Encouraging `for` loops discourages chaining and encourages single-iteration logic (e.g. using a `continue` instead of `filter`).
+ - `for` loops are considered "more readable" and have **clearer intent**.
+ - `for...of` loops offer the **most flexibility** for iteration (especially vs `Array.from`).
+
+For more detail, here is a break down of each of those points:
+
+### Layering of bad practices
+
+Typically developers will reach for a `forEach` when they want to iterate over a set of items. However not all "iterables" have access to Array methods. So a developer might convert their iterable to an Array by using `Array.from(iter).forEach()`. This code has introduced performance problems, where a `for...of` loop would be more performant.
+
+`forEach` does not do anything special with the Array - it does not create a new array or does not aid in encapsulation (except for introducing a new lexical scope within the callback, which isn't a benefit considering we use `let`/`const`). We don't dissallow `map`/`filter`/`reduce` because they have a tangible effect - they create a new array - which would take _more_ code and be _less_ readable to do with a `for...of` loop, the exception being as more requirements are added, and we start chaining array methods together...
+
+### Chaining 
+
+Often when using a method like `forEach` - when coming back to add new code, let's say to filter certain elements from the Array before operating on them, a developer is implicitly encouraged to use Array's method chaining to achieve this result. For example if we wanted to filter out bad apples from an Array of Apples, if the code already uses `forEach`, then its a simple addition to add `filter()`:
+
+```diff
+ apples
++  .filter(apple => !apple.bad)
+    .forEach(polishApple)
+```
+
+The problem we now have is that we're iterating multiple times over the items in a collection. Using `forEach` to begin with is what encouraged the chaining, if this were a `for` loop then the equivalent behaviour would be to use 2 `for` loops, which a developer is far less likely to write, so the `for` loop instead encourages an imperative style `continue`, keeping within a single set of iterations:
```

Comment:
I don't use forEach because it's purely for side effects, so a for loop avoids creating a closure and advertises it's just doing something for the side effect.

However, I prefer the functional style afforded by map and filter and will happily chain them. This avoids the goto behavior of `continue` and `next` and advertises a series of transformations that are independent of one another. 

Coming back to chained code later, it's easy to insert and remove new transformations. With a for loop, all intent is mixed together and cannot be easily changed.

Reply:
How do you suggest we change this paragraph to reflect that?

Reply:
I don't have an issue with the advice, just noting that personal preference will vary here and loop control flow is a sign that the loop is responsible for more than one thing.

A for loop that needs to both filter and map items can be structured like this to avoid control flow.

```js
function mapItems(items) {
  const results = []
  for (const item of items) {
    if (isSomePredicate(item) {
      results.push(doSomething(item))
    }
  }
  return results
}
const results = mapItems(items)
```

Pulling the loop body into a named function documents what it does and isolates its local variable state from the surrounding loop code. So it has the same benefits as my personal preference for:

```js
const results = items.filter(isSomePredicate).map(doSomething)
```

Reply:
Not a blocker for shipping this PR, but I just want to echo @dgraham's point that functional-style chaining "advertises a series of transformations that are independent of one another". In general, I think the readability of this chaining is often worth the (usually small) performance impact of doing numerous loops.

Reply:
I've updated to add some clarifications based on the comments.